src/videos/src/videos-service/app.py

```diff
@@ -0,0 +1,339 @@
+    """
+    try:
+        response = ssm_client.get_parameter(Name=parameter_name)
+        return response['Parameter']['Value'] != 'NONE'
+    except ssm_client.exceptions.ParameterNotFound:
+        return False
+
+
+def put_ivs_metadata(channel_arn, line):
+    """
+        Sends metadata to a given IVS stream. Metadata can be any string, but the AWS Retail Demo Store UI expects
+        metadata of the format {"productId":"<product-id>"}
+    """
+#     while True:
+    try:
+        app.logger.info(f'Sending metadata to stream: {line}')
+        ivs_client.put_metadata(
+            channelArn=channel_arn,
+            metadata=line
+        )
+    except ivs_client.exceptions.ChannelNotBroadcasting:
+        app.logger.warning(f'Channel not broadcasting. Waiting for 5 seconds.')
+        app.logger.info('Running ffmpeg processes:')
+        app.logger.info(os.system("ps aux|grep 'PID\|ffmpeg'"))
+        time.sleep(5)
+
+
+def get_stream_state(channel_arn):
+    """
+        Returns the state of a stream given it's ARN. One of 'LIVE', 'OFFLINE' (from API response)
+        or 'NOT_BROADCASTING' (inferred).
+    """
+    try:
+        stream_response = ivs_client.get_stream(channelArn=channel_arn)['stream']
+        stream_state = stream_response['state']
+    except ivs_client.exceptions.ChannelNotBroadcasting:
+        stream_state = "NOT_BROADCASTING"
+    return stream_state
+
+
+def start_streams():
+    """
+        Initiates all IVS streams based on environment variables. If the SSM_VIDEO_CHANNEL_MAP_PARAM (map of videos in
+        S3 to IVS channels) is set and the user has not requested to use the default IVS streams
+        (USE_DEFAULT_IVS_STREAMS, defined by CloudFormation input) then one stream will be started per video described
+        in the video to IVS channel map. Each stream runs in a separate thread.
+
+        If streams are not started, then `stream_details` will be set to the details of a collection of existing streams
+    """
+    if is_ssm_parameter_set(SSM_VIDEO_CHANNEL_MAP_PARAM) and not USE_DEFAULT_IVS_STREAMS:
+        video_channel_param_value = ssm_client.get_parameter(Name=SSM_VIDEO_CHANNEL_MAP_PARAM)['Parameter']['Value']
+        app.logger.info(f"Found IVS channel map: {video_channel_param_value}")
+        video_channel_map = json.loads(video_channel_param_value)
+
+        for idx, (s3_video_key, ivs_channel_arn) in enumerate(video_channel_map.items()):
+            threading.Thread(target=stream, args=(s3_video_key, ivs_channel_arn, idx)).start()
+
+    else:
+        global stream_details
+        stream_details = DEFAULT_STREAM_DETAILS
+
+
+def stream(s3_video_key, ivs_channel_arn, channel_id):
+    """
+        Starts the stream for a given video file and IVS channel. The video file is streamed on a loop using ffmpeg, and
+        any attached metadata (from the subtitles embedded in the video file) is sent to the channel's `put_metadata`
+        endpoint.
+    """
```

Comment:
So we are streaming from this Docker container? Doesn't IVS do the streaming? Sorry for the naive question.

Reply:
IVS provides the channel from which video content can be consumed, but the video must be streamed to the channel from an external source (IVS is the middle-man that allows us to scale to N subscribers along with its other features). This is either using live-streaming software which a user would run as part of a broadcast setup (beyond our scope) or using pre-recorded video. Our implementation is based on the ‘Streaming a Recorded Video with FFmpeg’ section of this guide: https://docs.aws.amazon.com/ivs/latest/userguide/GSIVS-live-streaming.html